pages/gall/[id]/index.tsx

```diff
@@ -68,6 +99,19 @@ const Gall = ({ species, imagePaths }: Props): JSX.Element => {
         setSelectedSource(s);
     };
 
+    const sortSourceList = () => {
+        const newSort = sourceSortOptions[(sourceList.sort.id + 1) % sourceSortOptions.length];
+        const sortOrder = newSort.order;
+        const sortProperty = newSort.property;
+
+        const sortedData = [...sourceList.data];
+        sortedData.sort((a,b) => sortOrder * a.source[sortProperty].localeCompare(b.source[sortProperty]));
```

Comment:
This line is not type safe since SourceApi can not be indexed by an arbitrary string value. The TypeScript way to handle this is to do this:

```typescript
type SortOption = {
    id: number;
    text: string;
    order: 1 | -1;
    property: keyof Pick<SourceApi, 'pubyear' | 'citation'>;
};

const sourceSortOptions: SortOption[] = [
    // as before
];
```
See https://www.typescriptlang.org/docs/handbook/utility-types.html#picktype-keys for info about the Pick higher order type.

`keyof` is a Typescript operator that defines a type based on the properties of an object.

This provides type safety on the SortOption type and allows you to index the SourceApi object in a type safe way. As a bonus we also get type safety on the sort order.

Reply:
Oh this is really cool!
I think it would be even better factored as:

```ts
type SortPropertyOption = {
    property: keyof Pick<SourceApi, 'pubyear' | 'citation'>;
    propertyText: string;
};

type SortOrderOption = {
     order: 1 | -1;
};

const sourceSortPropertyOptions: SortPropertyOption[] = [
    // only property and propertyText
];
```
and let the hook be:
```
{
    data,        // the array of `speciessource`
    sortIndex    // integer index into `sourceSortPropertyOptions`
    sortOrder    // SortOrderOption
}
```
This reduces the redundancy and allows us to change each piece individually.